### ISSUE TITLE
MSS310 - Sign error

### ISSUE THREAD

**Issue by dragon2611** [skullydazed/meross_mqtt_ha#1]
I managed to get the MSS310 onto the local broker, and it looks like it was discovered in Home assistant but when trying to switch it the following error is returned on the MQTT broker, I don't believe I have a shared key set, mss310 was configured for the broker using merros_powermon.

According to the payload from the device on connecting its firmware version is 2.1.16

```
{
  "header" : {
    "from" : "/appliance/controller",
    "method" : "SET",
    "namespace" : "Appliance.Control.ToggleX",
    "payloadVersion" : 1,
    "timestamp" : 1609167097,
    "messageId" : "f0e38cb19bad4ea79228e031f48b863d",
    "sign" : "0a9d7865a72e7cc5c3123fef469d699e"
  },
  "payload" : {
    "togglex" : {
      "channel" : 0,
      "onoff" : 1
    }
  }
}

```
```
{
  "header" : {
    "messageId" : "f0e38cb19bad4ea79228e031f48b863d",
    "namespace" : "Appliance.Control.ToggleX",
    "method" : "ERROR",
    "payloadVersion" : 1,
    "from" : "/appliance/myuuid/publish",
    "timestamp" : 1609167097,
    "timestampMs" : 805,
    "sign" : "b60e93b0a34ff2de6bfc809668050101"
  },
  "payload" : {
    "error" : {
      "code" : 5001,
      "detail" : "sign error"
    }
  }
}
```

Any ideas on how to resolve?

**Comment by skullydazed** [comment]
I believe you have a shared key set, the signature from the device does not validate correctly for a device with an empty shared key:

```
>>> from hashlib import md5
>>> sign = "f0e38cb19bad4ea79228e031f48b863d" + "1609167097"
>>> md5(sign.encode('utf-8')).hexdigest()
'0a9d7865a72e7cc5c3123fef469d699e'
```

**Comment by dragon2611** [comment]
I don't believe I've set one, so it's either remembered it from when it got connected to their cloud or I did set one and forgot, although that said I don't see any such options in the meross_powermon I'd used to configure it to the local broker.

Do you know if there's any way to reset/overwrite it?

**Comment by skullydazed** [comment]
I'm not familiar with meross_powermon so I can't help with how that handles pairing.

I used https://github.com/bytespider/Meross for the pairing process. If you clone that somewhere you can use `bin/src/meross-info` to query the plugs over HTTP. If no key is set you should get back information like this:

```
$ ./meross info -g 172.16.25.31
Getting info about device with IP 172.16.25.31
sending payload {
  header: {
    method: 'GET',
    namespace: 'Appliance.System.All',
    messageId: '6ac3606f592bae270a1b8339a1905dc6',
    timestamp: 1609172291,
    sign: 'ee46b278a833eb74c7706303abae17f9'
  },
  payload: {}
}
Got response: {
  header: {
    messageId: '6ac3606f592bae270a1b8339a1905dc6',
    namespace: 'Appliance.System.All',
    method: 'GETACK',
    payloadVersion: 1,
    from: '/appliance/1909181776048925186548e1e9054d19/publish',
    timestamp: 1609172290,
    timestampMs: 124,
    sign: '11076311b728e0006789402704a7cdb5'
  },
  payload: {
    all: {
      system: {
        hardware: {
          type: 'mss110',
          subType: 'us',
          version: '2.0.0',
          chipType: 'mt7682',
          uuid: '1909181776048925186548e1e9054d19',
          macAddress: '48:e1:e9:05:4d:19'
        },
        firmware: {
          version: '2.1.19',
          compileTime: '2020/04/15 17:02:04 GMT +08:00',
          wifiMac: '0c:80:63:dd:c5:6b',
          innerIp: '172.16.25.31',
          server: '172.16.25.1',
          port: 8883,
          userId: 0
        },
        time: { timestamp: 1609172290, timezone: '', timeRule: [] },
        online: { status: 1 }
      },
      digest: {
        togglex: [ { channel: 0, onoff: 0, lmTime: 1608569273 } ],
        triggerx: [],
        timerx: []
      }
    }
  }
}
```

**Comment by dragon2611** [comment]
Thanks, I reset the plug and paired it using the utility you linked it's working now.

I don't suppose you know how to read the power usage data (Not really important, just a nice to have)

**Comment by skullydazed** [comment]
My plugs don't report power usage, and for future plugs I moved to esphome anyway. If they don't report periodically to mqtt it [looks like you can send an "Appliance.Control.Electricity" request](https://github.com/davidb24v/meross-powermon/blob/master/meross_powermon/modified_device.py#L105). I'm afraid it'll be a while before something like this could filter up to the top of my todo list, but if you or someone else want to implement that it wouldn't be too hard to replace the paho loop_forever() call with a while loop that periodically sends that request.

**Comment by Pierremenard** [comment]
> My plugs don't report power usage, and for future plugs I moved to esphome anyway. If they don't report periodically to mqtt it [looks like you can send an "Appliance.Control.Electricity" request](https://github.com/davidb24v/meross-powermon/blob/master/meross_powermon/modified_device.py#L105). I'm afraid it'll be a while before something like this could filter up to the top of my todo list, but if you or someone else want to implement that it wouldn't be too hard to replace the paho loop_forever() call with a while loop that periodically sends that request.

When I send this kind of request in the local MQTT server I get an answer, but the reported power is always 0. I thought it was a device failure, but resetting it and using with the original Meross app the power is correct.

My request is:

{
  "header": {
    "from": "/app/6549264-3512a18a5hf5c7354d00f250d5bd09ba/subscribe",
    "messageId": "9f97e7907344cea2737abd6fb965d786",
    "method": "GET",
    "namespace": "Appliance.Control.Electricity",
    "payloadVersion": 1,
    "sign": "28a7a661c4e70cc0e939eac20ad2fc5a",
    "timestamp": 57450
  },
  "payload": {
    "channel": 0
  }
}

and answer is

{
  "header": {
    "messageId": "9f97e7907344cea2737abd6fb965d786",
    "namespace": "Appliance.Control.Electricity",
    "method": "GETACK",
    "payloadVersion": 1,
    "from": "/appliance/2104020442298751806048e1e969b15c/publish",
    "timestamp": 3802,
    "timestampMs": 509,
    "sign": "57b2f456d13e47d17a76b07d88e52049"
  },
  "payload": {
    "electricity": {
      "channel": 0,
      "current": 0,
      "voltage": 0,
      "power": 0,
      "config": {
        "voltageRatio": 188,
        "electricityRatio": 102
      }
    }
  }
}
